# 20260226 — Vendor Portal — Implementation Plan

> Source methodology: `docs/methodologies/20260226.VendorPortal.ShipmentFlow.md`
>
> This document is the ordered, file-level implementation plan for bringing the Vendor Portal into compliance with CoreWeave logistics requirements. Each phase is self-contained, buildable, and testable before moving on.

---

## Phase 0 — Dependencies

**Goal:** Install third-party libraries needed by later phases.

### 0.1 New npm packages
| Package | Purpose |
|---|---|
| `react-barcode` | Code 128 barcode rendering for unit/carton serial labels (Section 5 of CoreWeave doc) |
| `jspdf` | Client-side PDF generation for packing slip, packing list, commercial invoice |
| `jspdf-autotable` | Table layout plugin for jspdf |

### 0.2 Action
```
npm install react-barcode jspdf jspdf-autotable
```

### 0.3 Rule
- Do not upgrade or change any existing dependency versions.
- All new packages must be added to `dependencies` (not `devDependencies`).

---

## Phase 1 — Data Model & State

**Goal:** Extend the data model so shipments can carry all CoreWeave-required fields without breaking existing Coreweave-side functionality.

### 1.1 Modify `src/utils/storage.js`
- The `loadDatabase` default shape must add a `commercialInvoices: []` array alongside the existing `purchaseOrders`, `shipments`, `receivings`.
- `importDatabaseFromJSON` validation must accept the new `commercialInvoices` key (treat as optional for backward compat with old exports).

### 1.2 Modify `src/context/AppContext.jsx` — `CREATE_SHIPMENT` action
Extend the shipment object created inside the reducer. The new shape:

```
{
  id,
  poIds: [],                  // array — replaces singular poId for multi-PO support
  poId,                       // keep for backward compat; set to poIds[0]
  shippedAt,
  shipmentType: 'domestic' | 'international',
  incoterms: '' | null,
  trackingNumber,
  notes,
  pallets: [
    { id, palletId, type: 'single-sku' | 'mixed-sku' }
  ],
  packedUnits: [
    { label, palletId?, lengthVal, widthVal, heightVal, dimUnit: 'in'|'cm',
      weightVal, weightUnit: 'lbs'|'kg', contents: [{ skuDescription, qty }] }
  ],
  items: [
    { id, poItemId, poId, quantityShipped, serialsShipped, mpn, palletId? }
  ]
}
```

#### Rules
- Existing `poId` field must remain populated (set to first entry in `poIds`) so Receiving and Reconciliation pages continue to work without changes.
- Every field added to shipment items (`mpn`, `palletId`) defaults to `null` so old code paths are unaffected.
- Add `shipmentType` default `'domestic'` so the existing Coreweave-side "Vendor Portal" view (used by internal staff) keeps working as-is.
- The reducer must also handle multi-PO status transitions: iterate `poIds` and set each PO to `shipped`.

### 1.3 Add new reducer action `CREATE_COMMERCIAL_INVOICE`
Shape:
```
{
  id,
  shipmentId,
  createdAt,
  exporter: { name, address, contact },
  consignee: { name, address, contact },
  invoiceNumber,
  invoiceDate,
  currency,
  incoterms,
  reasonForExport,
  lineItems: [
    { sku, mpn, description, hsCode, countryOfOrigin,
      quantity, unitOfMeasure, unitPrice, extendedValue }
  ],
  subtotal,
  freightInsurance,
  grandTotal
}
```

Store in `state.commercialInvoices`.

### 1.4 Rule — No breaking changes
- All existing pages (`Dashboard`, `PurchaseOrders`, `CreatePO`, `PODetail`, `Receiving`, `Reconciliation`, `DataManagement`) must remain fully functional after this phase. Run the app and verify before proceeding.

---

## Phase 2 — Vendor-Scoped PO Filtering

**Goal:** When a vendor user is logged in, they only see POs that belong to their company.

### 2.1 Modify `src/pages/VendorPortal.jsx`
- Import `useAuth` from `../context/AuthContext`.
- The `availablePOs` filter (currently lines 20-22) must add a vendor match condition:
  ```
  const { user } = useAuth();
  const vendorName = user?.vendor?.shortName || user?.vendor?.name || '';
  const availablePOs = state.purchaseOrders.filter(
    (po) =>
      (po.status === 'issued' || po.status === 'shipped') &&
      (!vendorName || po.vendor.toLowerCase().includes(vendorName.toLowerCase()))
  );
  ```
- Use case-insensitive partial match because PO vendor names are free-text entered by Coreweave staff and may not exactly match vendor `shortName`.

### 2.2 Rule — Coreweave users see all
- When `user.role === 'coreweave'`, the vendor filter must be bypassed (vendorName will be empty, so the `!vendorName` guard handles this).

---

## Phase 3 — Multi-Step Vendor Shipment Wizard

**Goal:** Replace the current single-form VendorPortal with a stepped wizard matching the methodology flow.

### 3.1 Create `src/pages/vendor/ShipmentWizard.jsx`
This is the new top-level vendor page component. It manages a `step` state (0-4) and renders one sub-component per step:

| Step | Component | Description |
|---|---|---|
| 0 | `POSelection` | Multi-select PO table filtered by vendor |
| 1 | `ShipmentInfo` | Carrier, tracking, shipment type, pallet registration |
| 2 | `LineItemEntry` | Per-item SKU, MPN, qty, serials, pallet assignment |
| 3 | `PackedDimensions` | Dimensions & weight per packed unit |
| 4 | `ReviewSubmit` | Consolidated review, validation gate, submit |

Wizard state is held in a single `shipmentDraft` object at the wizard level and passed down via props. Each step calls a `updateDraft(partialData)` callback to merge its fields.

### 3.2 Create `src/pages/vendor/POSelection.jsx`

#### Layout
- Card with table: checkbox column, PO Number, Vendor, Status, Created, Items count.
- Only POs matching the logged-in vendor and in `issued`/`shipped` status are shown.
- Multi-select via checkboxes.

#### Rules
- At least one PO must be selected to proceed.
- "Next" button disabled until selection is non-empty.
- On "Next", populate `shipmentDraft.poIds` and pre-populate line items from all selected POs.

### 3.3 Create `src/pages/vendor/ShipmentInfo.jsx`

#### Layout
- Card with form fields: Tracking Number (text), Shipment Type (radio: Domestic / International), Incoterms (text, conditionally shown), Notes (textarea).
- Pallet Registration sub-section: "Add Pallet" button that appends a row with Pallet ID (text) and Type (select: single-sku / mixed-sku). Each row has a remove button.

#### Rules
- Tracking Number is required.
- Shipment Type defaults to Domestic.
- Incoterms field appears only when International is selected and is then required.
- Pallet ID must be unique within the shipment; show inline validation error on duplicates.
- Pallet registration is optional for single-pallet shipments.

### 3.4 Create `src/pages/vendor/LineItemEntry.jsx`

#### Layout
- One card per PO, with PO Number as the card heading.
- Inside each card, one `.line-item` block per PO item containing:
  - **CoreWeave Friendly SKU** — read-only text from `item.description`.
  - **MPN** — pre-populated text input from `item.partNumber`; editable; shows "N/A" placeholder if empty.
  - **Quantity Shipped** — number input, min 1, max `quantityOrdered`.
  - **Serial Numbers** — existing serial input + tag pattern (reuse current UX).
  - **Pallet Assignment** — select dropdown listing registered Pallet IDs (from Step 1), or "None" if no pallets were registered. Only visible if pallets exist.

#### Rules
- Quantity must be > 0 and ≤ ordered quantity.
- Serial count must equal Quantity Shipped. Show a warning badge if mismatched; block "Next" if any item fails.
- Serial numbers must be globally unique across all items in the shipment. Highlight duplicates in red (`serial-tag tag-danger`).
- If pallets are registered, every line item must be assigned to a pallet. Block "Next" if unassigned.

### 3.5 Create `src/pages/vendor/PackedDimensions.jsx`

#### Layout
- If pallets were registered, pre-populate one row per pallet (label = Pallet ID). Otherwise start with one empty "Packed Unit" row.
- "Add Packed Unit" button for additional cartons/crates.
- Each row:
  - Label (text input, pre-filled with Pallet ID if applicable)
  - Length, Width, Height (number inputs) + Unit selector (in / cm)
  - Weight (number input) + Unit selector (lbs / kg)
  - Contents — auto-generated read-only summary (SKUs × qty assigned to that pallet/unit)
- Remove button per row (minimum 1 row must remain).

#### Rules
- All dimension and weight fields are required and must be > 0.
- At least one packed unit is required.
- Unit selectors default to inches and lbs.

### 3.6 Create `src/pages/vendor/ReviewSubmit.jsx`

#### Layout
Five collapsible sections:

1. **Shipment Summary** — read-only display of carrier, tracking, type, incoterms.
2. **PO Line Items** — table grouped by PO Number; columns: SKU, MPN, Qty Shipped, Serials, Pallet.
3. **Packed Dimensions & Weight** — table: Label, L×W×H, Weight, Contents.
4. **Commercial Invoice** (shown only if international) — see Phase 5; for now show "International invoice section — coming in Phase 5" placeholder.
5. **Validation Status** — list of checks with green checkmark or red X per rule.

#### Validation gate (runs on mount and on any draft change)
| Check | Blocks submit? |
|---|---|
| All items have qty > 0 and ≤ ordered | Yes |
| Serial count = qty for each item | Yes |
| No duplicate serials | Yes |
| All items assigned to a pallet (if pallets registered) | Yes |
| All packed units have dimensions & weight > 0 | Yes |
| At least one packed unit exists | Yes |
| Tracking number is non-empty | Yes |
| Incoterms filled (if international) | Yes |

"Submit Shipment" button is disabled until all checks pass.

#### On submit
- Dispatch `CREATE_SHIPMENT` with the full draft.
- Transition to **Post-Submission** screen (Phase 4).

### 3.7 Update routing

#### `src/App.jsx`
- Import `ShipmentWizard`.
- In the vendor `ProtectedRoutes`, replace the single `VendorPortal` route with:
  ```
  <Route index element={<ShipmentWizard />} />
  <Route path="vendor-portal" element={<ShipmentWizard />} />
  <Route path="shipment-history" element={<VendorPortal />} />
  ```
- Keep the old `VendorPortal.jsx` at `/shipment-history` as a read-only history view (it still works for viewing past shipments from the Coreweave side too).

#### `src/components/Layout.jsx`
- In `vendorNav`, update:
  ```
  { to: '/', icon: Truck, label: 'New Shipment' },
  { to: '/shipment-history', icon: ClipboardList, label: 'History' },
  ```

### 3.8 CSS additions — `src/index.css`
- `.wizard-steps` — horizontal step indicator bar (step number + label, active/completed/pending states).
- `.wizard-nav` — flex row with Back / Next buttons at the bottom of each step.
- `.wizard-step-content` — wrapper with the `fadeIn` animation.
- `.validation-list` — list of validation results with `.check-pass` (green) and `.check-fail` (red) items.
- `.collapsible-section` — card section with toggle header.

#### Rules
- Reuse existing design tokens (`--primary`, `--success`, `--danger`, `--border`, `--radius`, `--shadow`).
- Reuse existing component classes (`.card`, `.form-stack`, `.form-group`, `.form-row`, `.btn`, `.serial-tag`, `.line-item`).
- Do not modify any existing CSS class definitions; only add new ones.

---

## Phase 4 — Post-Submission Outputs

**Goal:** After successful submission, display and offer printable/downloadable outputs.

### 4.1 Create `src/pages/vendor/ShipmentConfirmation.jsx`
Rendered by `ShipmentWizard` after successful submit. Shows:

1. **Success banner** — reuse `.success-card` pattern.
2. **Pallet QR codes** — one `<QRCodeSVG>` per single-SKU pallet encoding: `{ palletId, poNumber, sku, mpn, quantity, serialNumbers }`.
3. **Unit/carton barcode labels** — one `<Barcode>` (from `react-barcode`) per serial number. Displayed in a print-friendly grid.
4. **Action buttons:**
   - Print Packing Slip — calls `generatePackingSlipPDF()`.
   - Print Packing List — calls `generatePackingListPDF()`.
   - Print QR Labels — `window.print()` scoped to QR section.
   - Print Barcode Labels — `window.print()` scoped to barcode section.
   - Record Another Shipment — resets wizard.

### 4.2 Create `src/utils/pdfGenerators.js`
Export three functions:

#### `generatePackingSlipPDF(shipmentData, purchaseOrders)`
Builds a PDF with:
- Header: "Packing Slip" + date.
- Shipment-level: PO Number(s), Vendor, Tracking Number.
- Per PO: table of line items (SKU, MPN, Qty Shipped, Serial Numbers).
- Pallet ID references where applicable.

#### `generatePackingListPDF(shipmentData)`
Builds a PDF with:
- All packing slip content, plus:
- Packed Dimensions & Weight table (Label, L×W×H, Unit, Weight, Unit, Contents).

#### `generateCommercialInvoicePDF(invoice)`
Builds a PDF with:
- Exporter / Consignee block.
- Invoice metadata (number, date, currency, incoterms, reason).
- Line-item table (SKU, MPN, Description, HS Code, Origin, Qty, Unit Price, Ext. Value).
- Totals block.

#### Rules
- Use `jspdf` + `jspdf-autotable` for all PDF generation.
- All PDFs must include the text "CoreWeave Purchase Reconciliation" in the footer.
- Use A4 page size with 15mm margins.

### 4.3 Print stylesheet additions — `src/index.css`
- Add `@media print` rules for `.qr-label-grid` and `.barcode-label-grid` to ensure clean printing (hide non-label content, page breaks between pallets).

---

## Phase 5 — Commercial Invoice (International Shipments)

**Goal:** Add the commercial invoice form and PDF generation for international shipments.

### 5.1 Create `src/pages/vendor/CommercialInvoice.jsx`
This is rendered as a sub-step inside `ReviewSubmit` (or as an additional wizard step inserted before Review) when `shipmentType === 'international'`.

#### Layout
- **Header fields card:** Exporter (pre-filled from vendor data), Consignee (pre-filled from PO), Invoice Number, Invoice Date, Currency selector, Incoterms (carried from Step 1), Reason for Export selector.
- **Line items card:** table pre-populated from shipment line items. Additional columns for: Description of Goods (text), HS/Tariff Code (text, min 6 chars), Country of Origin (text), Unit Price (number), Extended Value (auto-calc).
- **Totals card:** Subtotal (auto), Freight/Insurance (number, optional), Grand Total (auto).

#### Rules
- All header fields are required.
- HS Code must be at least 6 digits.
- Extended Value = Quantity × Unit Price (auto-calculated, read-only).
- Grand Total = Subtotal + Freight/Insurance.
- Validation: totals on invoice must match line-item sum. Flag mismatches in the review screen.

### 5.2 Update `ReviewSubmit.jsx`
- Replace the Phase 3 placeholder with the actual `CommercialInvoice` component output (read-only preview mode).
- Add commercial invoice validation checks to the validation gate.

### 5.3 Update `ShipmentConfirmation.jsx`
- Add "Download Commercial Invoice" button calling `generateCommercialInvoicePDF()`.

### 5.4 Wire up `CREATE_COMMERCIAL_INVOICE` dispatch
- On shipment submit, if international, also dispatch `CREATE_COMMERCIAL_INVOICE` with the invoice data.

---

## Phase 6 — Validation Hardening & Edge Cases

**Goal:** Ensure robustness and compliance with every CoreWeave rule.

### 6.1 Serial number uniqueness — cross-shipment check
- On serial entry, check `state.shipments` to warn if a serial has been used in a prior shipment (informational warning, not a hard block — serials could be re-shipped for RMA).

### 6.2 Partial shipment tracking
- When a vendor ships less than the full quantity, the PO status should remain `issued` (not transition to `shipped`) so it appears again for a follow-up shipment.
- Add logic: if **every** line item's total shipped quantity (across all shipments) equals or exceeds ordered quantity, then set status to `shipped`.

### 6.3 Mixed-SKU pallet validation
- For pallets marked `mixed-sku`, the system must verify that multiple distinct SKUs are assigned. If only one SKU is assigned, show a warning suggesting the vendor change the pallet type to `single-sku` (which enables QR barcode generation).

### 6.4 Pallet QR character limit handling
- Per CoreWeave Section 4.3: "If character limits require, serial numbers may be encoded as a structured list or range." If the QR payload exceeds 2,953 bytes (QR alphanumeric limit), fall back to encoding serials as a range (e.g., `SN001-SN100`) or split across multiple QR codes per pallet. Display a notice to the vendor if splitting is necessary.

---

## Phase 7 — Polish & UX

**Goal:** Final pass for user experience consistency.

### 7.1 Wizard progress indicator
- Numbered step bar at the top of the wizard: `1 Select POs → 2 Shipment Info → 3 Line Items → 4 Dimensions → 5 Review`.
- Active step highlighted in `--primary`; completed steps show a green checkmark; future steps are muted.

### 7.2 Inline help text
- Add `text-muted` helper text below key fields referencing the CoreWeave requirement:
  - SKU field: "Must match the CoreWeave PO exactly."
  - MPN field: "Enter 'N/A' if not applicable."
  - Serial Numbers: "List every serial for this SKU. Count must match quantity shipped."
  - Pallet ID: "Unique alphanumeric ID — must match the physical pallet label."

### 7.3 Mobile responsiveness
- Wizard steps collapse to a compact indicator on screens < 768px.
- Form rows stack vertically (already handled by existing `.form-row` responsive rule).
- QR/barcode print layouts remain full-width.

### 7.4 Loading & error states
- Show a brief loading spinner during PDF generation.
- Show `alert-error` for any dispatch failures.

---

## File Inventory — Summary

### New files to create
| File | Phase |
|---|---|
| `src/pages/vendor/ShipmentWizard.jsx` | 3.1 |
| `src/pages/vendor/POSelection.jsx` | 3.2 |
| `src/pages/vendor/ShipmentInfo.jsx` | 3.3 |
| `src/pages/vendor/LineItemEntry.jsx` | 3.4 |
| `src/pages/vendor/PackedDimensions.jsx` | 3.5 |
| `src/pages/vendor/ReviewSubmit.jsx` | 3.6 |
| `src/pages/vendor/ShipmentConfirmation.jsx` | 4.1 |
| `src/pages/vendor/CommercialInvoice.jsx` | 5.1 |
| `src/utils/pdfGenerators.js` | 4.2 |

### Existing files to modify
| File | Phase | Changes |
|---|---|---|
| `package.json` | 0 | Add `react-barcode`, `jspdf`, `jspdf-autotable` |
| `src/utils/storage.js` | 1.1 | Add `commercialInvoices` to default shape and import validation |
| `src/context/AppContext.jsx` | 1.2, 1.3 | Extend `CREATE_SHIPMENT`, add `CREATE_COMMERCIAL_INVOICE` |
| `src/pages/VendorPortal.jsx` | 2.1 | Add vendor-scoped PO filter via `useAuth` |
| `src/App.jsx` | 3.7 | Add vendor wizard route, move old portal to `/shipment-history` |
| `src/components/Layout.jsx` | 3.7 | Update `vendorNav` entries |
| `src/index.css` | 3.8, 4.3 | Add wizard, validation, print styles |

### Files NOT modified (must remain unchanged)
| File | Reason |
|---|---|
| `src/pages/Dashboard.jsx` | Coreweave-only; no vendor changes |
| `src/pages/PurchaseOrders.jsx` | Coreweave-only |
| `src/pages/CreatePO.jsx` | Coreweave-only |
| `src/pages/PODetail.jsx` | Coreweave-only |
| `src/pages/Receiving.jsx` | Coreweave-only |
| `src/pages/Reconciliation.jsx` | Coreweave-only |
| `src/pages/DataManagement.jsx` | Coreweave-only |
| `src/utils/reconcile.js` | Reconciliation logic; vendor changes must not alter |
| `src/context/AuthContext.jsx` | Already complete from prior work |
| `src/pages/Login.jsx` | Already complete from prior work |
| `src/data/cores/vendors/*.ts` | Already complete from prior work |

---

## Execution Order

```
Phase 0  →  npm install new deps
Phase 1  →  Data model & state (no UI changes)
Phase 2  →  Vendor-scoped filtering (small, testable)
Phase 3  →  Wizard UI (largest phase — build & test each step incrementally)
Phase 4  →  Post-submission outputs (QR, barcodes, PDFs)
Phase 5  →  Commercial invoice (international only)
Phase 6  →  Validation hardening & edge cases
Phase 7  →  Polish & UX
```

Each phase should end with a successful `npm run build` and manual smoke test. Do not begin the next phase until the current phase builds and runs cleanly.

---

## Guiding Rules (Apply to ALL Phases)

1. **Backward compatibility** — Existing Coreweave-side pages must not break. Every data model extension must default gracefully.
2. **Existing CSS reuse** — Use existing design tokens and component classes. Only add new CSS classes; never modify existing ones.
3. **No TypeScript conversion** — New page and utility files must be `.jsx` / `.js` to match the existing codebase. The `.ts` files in `src/data/cores/vendors/` are the only TypeScript in the project; Vite handles them transparently.
4. **Component patterns** — Follow the existing page pattern: `export default function ComponentName()`, use `useApp()` for state, `useAuth()` for user, destructure from `lucide-react` for icons.
5. **Form patterns** — Use `.form-stack`, `.form-group`, `.form-row`, `.form-actions` classes. Controlled inputs with React state.
6. **No external API calls** — All data lives in localStorage via `AppContext`. No backend required.
7. **Build verification** — Run `npm run build` after every phase. Zero errors, zero warnings.
